core/src/main/java/software/amazon/awssdk/retry/RetryPolicy.java

```diff
@@ -15,222 +15,106 @@
-
-    /**
-     * Back-off strategy to control the sleep time between retry attempts. This
-     * field should not be null.
-     */
-    private final BackoffStrategy backoffStrategy;
-
-    /**
-     * Non-negative integer indicating the max retry count.
-     */
-    private final int maxErrorRetry;
-
-    /**
-     * Whether this retry policy should honor the max error retry set in ClientConfiguration.
-     */
-    private final boolean honorMaxErrorRetryInClientConfig;
-
-    /**
-     * Constructs a new retry policy. See {@link PredefinedRetryPolicies} for
-     * some pre-defined policy components, and also the default policies used by
-     * SDK.
-     *
-     * @param retryCondition
-     *            Retry condition on whether a specific request and exception
-     *            should be retried. If null value is specified, the SDK'
-     *            default retry condition is used.
-     * @param backoffStrategy
-     *            Back-off strategy for controlling how long the next retry
-     *            should wait. If null value is specified, the SDK' default
-     *            exponential back-off strategy is used.
-     * @param maxErrorRetry
-     *            Maximum number of retry attempts for failed requests.
-     * @param honorMaxErrorRetryInClientConfig
-     *            Whether this retry policy should honor the max error retry set
-     *            in configuration
-     * @see PredefinedRetryPolicies
-     */
-    public RetryPolicy(RetryCondition retryCondition,
-                       BackoffStrategy backoffStrategy,
-                       int maxErrorRetry,
-                       boolean honorMaxErrorRetryInClientConfig) {
-        if (retryCondition == null) {
-            retryCondition = PredefinedRetryPolicies.DEFAULT_RETRY_CONDITION;
-        }
-        if (backoffStrategy == null) {
-            backoffStrategy = PredefinedRetryPolicies.DEFAULT_BACKOFF_STRATEGY;
-        }
-        if (maxErrorRetry < 0) {
-            throw new IllegalArgumentException("Please provide a non-negative value for maxErrorRetry.");
-        }
-
-        this.retryCondition = retryCondition;
-        this.backoffStrategy = backoffStrategy;
-        this.maxErrorRetry = maxErrorRetry;
-        this.honorMaxErrorRetryInClientConfig = honorMaxErrorRetryInClientConfig;
-    }
-
-    ;
-
-    /**
-     * Returns the retry condition included in this retry policy.
-     *
-     * @return The retry condition included in this retry policy.
-     */
-    public RetryCondition getRetryCondition() {
-        return retryCondition;
-    }
-
-    /**
-     * Returns the back-off strategy included in this retry policy.
-     *
-     * @return The back-off strategy included in this retry policy.
-     */
-    public BackoffStrategy getBackoffStrategy() {
-        return backoffStrategy;
+@SdkPublicApi
+public interface RetryPolicy extends BackoffStrategy, RetryCondition, ToCopyableBuilder<RetryPolicy.Builder, RetryPolicy> {
```

Comment:
It seems weird that a retry policy is-a backoff strategy and a retry condition. That seems like a has-a relationship. Any particular reason to extend the interfaces here?

Reply:
I like it extending the interfaces, otherwise we have to write awkward code in the runtime to get the individual strategies. Also let's you implement the condition and backoff strategy easily in one class.

Reply:
We could always implement both in the same class if we wanted by making the subclass implement the condition and backoff strategy interfaces. What's the awkward code we'd have to write in the runtime?

Here's my complaint about the way it currently works:

Given:
```Java
class MyRetryPolicy implements RetryPolicy {
...
}
```

We can create a retry policy that delegates to other retry policies, which to me is a bit odd:

```Java
RetryPolicy myRetryStrategy = new MyRetryPolicy();
RetryPolicy letsGetMeta = RetryPolicy.builder()
    .retryCondition(myRetryStrategy)
    .backoffStrategy(myRetryStrategy)
    .build();
```

Even if that's considered a "feature", it implies a retry policy must always delegate to a retry condition and backoff strategy, but it also IS a retry condition and backoff strategy? That has really weird implications.

For example, I can't even wrap my head around what this code does:

```Java
RetryPolicy whatDoesThisEvenDoPolicy = myRetryStrategy.toBuilder()
    .retryCondition(myRetryStrategy)
    .backoffStrategy(myRetryStrategy)
    .build();
```

`MyRetryPolicy` must implement `toBuilder()`,  because it extends `ToCopyableBuilder`, but it also must HAVE a condition and a backoff strategy because it can be converted to and from a builder that HAS one? So `RetryPolicy` always delegates to a `RetryCondition` and a `BackoffStrategy`?

My only guess at how to implement by own retry policy is to either throw a `UnsupportedOperationException` from `toBuilder` (meaning we've given up on the L in SOLID) or just make my retry policy always delegate to a condition and backoff strategy, at which point every RetryPolicy is a decorator (meaning we've given up on the I in SOLID). We will have already given up on the S in SOLID by having the class implement both interfaces.

Reply:
> So RetryPolicy always delegates to a RetryCondition and a BackoffStrategy?

Not necesarily. The simple implementation takes a condition and backoff strategy separately and delegates to them but you could just as easily implement it all in one class and combine the logic. I think the weirdness here stems from forcing the builder on the retry policy interface, do we even need a builder? Would a factory method suffice for the common case of combining condition and strategy?

Reply:
How does it break SOLID?

Reply:
S: We're defining an interface `RetryPolicy` with two potential responsibilities: defining a condition and defining a backoff strategy.
I: We have an interface `RetryPolicy` that has two methods that can (and already are in this case) be separated into two separate interfaces.

Reply:
I don't really agree with all that. In certainly can be broken up into two separate concepts but from a customer's perspective its one logical thing, how the client handles retry situations. Does any aggregate interface break SOLID? I wouldn't think so.